File: src/camera.ts
declare const cam_x: HTMLSpanElement;
declare const cam_y: HTMLSpanElement;
declare const cam_z: HTMLSpanElement;

/** Shared pan/zoom state. Both renderers read this to place world-space
 * draw commands on screen; input handling reads/writes it directly. */
export const camera = (() => {
  let x = 0;
  let y = 0;
  let zoom = 1;

  // prettier-ignore
  return {
    get x() { return x },
    get y() { return y },
    get zoom() { return zoom },

    set x(v:number) {
      x = v;
      cam_x.textContent = v.toFixed(2);
    },
    set y(v:number) {
      y = v;
      cam_y.textContent = v.toFixed(2);
    },
    set zoom(v: number) {
      zoom = v;
      cam_z.textContent = v.toFixed(2);
    },
  };
})();

export const MIN_ZOOM = 0.25;
export const MAX_ZOOM = 4;

export function toWorld(screenX: number, screenY: number): [f32, f32] {
  return [
    ((screenX - camera.x) / camera.zoom) as f32,
    ((screenY - camera.y) / camera.zoom) as f32,
  ];
}

/** Zoom in/out by `deltaY` (wheel units) while keeping the world point under
 * (screenX, screenY) fixed on screen. */
export function zoomAt(screenX: number, screenY: number, deltaY: number) {
  const [wx, wy] = toWorld(screenX, screenY);
  const newZoom = Math.min(
    MAX_ZOOM,
    Math.max(MIN_ZOOM, camera.zoom * Math.exp(-deltaY * 0.01)),
  );
  camera.x -= wx * (newZoom - camera.zoom);
  camera.y -= wy * (newZoom - camera.zoom);
  camera.zoom = newZoom;
}

export function pan(dx: number, dy: number) {
  camera.x -= dx;
  camera.y -= dy;
}

/** Direct drag-to-pan: the canvas follows the cursor 1:1 (opposite sign
 * convention from wheel-scroll's `pan`, which mimics natural scrolling). */
export function panByDrag(dx: number, dy: number) {
  camera.x += dx;
  camera.y += dy;
}

File: src/draw/camera.rs
use crate::ffi;

pub struct Camera {
    pub x: f32,
    pub y: f32,
    pub zoom: f32,
}

impl Camera {
    pub const MIN_ZOOM: f32 = 0.1;
    pub const MAX_ZOOM: f32 = 4.0;

    pub const fn new() -> Self {
        Self {
            x: 0.0,
            y: 0.0,
            zoom: 1.0,
        }
    }

    pub fn zoom_at(&mut self, sx: f32, sy: f32, dy: f32) {
        let (wx, wy) = self.to_world(sx, sy);
        let new_zoom = (self.zoom * ffi::expf(-dy * 0.01)).clamp(Self::MIN_ZOOM, Self::MAX_ZOOM);

        self.x = sx / new_zoom - wx;
        self.y = sy / new_zoom - wy;
        self.zoom = new_zoom;
    }

    pub fn to_world(&self, sx: f32, sy: f32) -> (f32, f32) {
        ((sx / self.zoom - self.x), (sy / self.zoom - self.y))
    }

    pub fn to_screen(&self, wx: f32, wy: f32) -> (f32, f32) {
        ((wx + self.x) * self.zoom, (wy + self.y) * self.zoom)
    }

    pub fn pan(&mut self, dx: f32, dy: f32) {
        self.x -= dx / self.zoom;
        self.y -= dy / self.zoom;
    }

    pub fn pan_by_drag(&mut self, dx: f32, dy: f32) {
        self.x += dx / self.zoom;
        self.y += dy / self.zoom;
    }
}
